### PR TITLE
feat: Include ontology details in search field

### PR DESCRIPTION
**Pull request by connoratrug**
Add the ontology ( ontology search values) to row search field

todo:

- [x] test and cleanup code, extract string hacking to separate function and add tests
- [x] migration to update existing triggers and indexes ( requires updating the search col value for each row in each table in each schema :( )

🚨 Need to evaluate performance ( specifically on migration ), consider removing migration and adding to change only for new schemas

### CODE REVIEW

**Comment by connoratrug** [comment]
example shows finding a cohort via collection event , ontology.code column value 
Screenshot 2023-09-27 at 16 43 34

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/TriggerBuilderHelpers.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+  private TriggerBuilderHelpers() {
+    throw new IllegalStateException("Utility class");
+  }
+
+  public static String buildOntologySearchSection(
+      String schemaName, String tableName, String columnName, boolean isArray) {
+    String template =
+        isArray
+            ? """
```
Comment (mswertz) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/TriggerBuilderHelpers.java:13]: Evil code ;-) you could have used jooq DSL builder for this. I don't mind this, jooq is evil in its own way.
So this comment is just FYI. That would be something like DSLContext.select() etc.

**Review by mswertz — COMMENTED** [review]
it now also includes 'parent' and 'child' because that is included in the ontology search right? is that too eager? 

Maybe this is more a comment about that we need to change the search index for ontology table than on this PR?

Meanwhile, I couldn't get it to work. => In pet store I annotated 'Tag=red' but then I didn't find results based on its label/code/etc

So I added this
image

And didn't find it
image

I suppose that is because I changed the 'color' label after the pet records were updated.
So then I edited one of the pets and saved.

image

 So we could accept this limitation but we might need to have some reindexing feature (hard) OR move the logic form the trigger to the SqlQuery -> search conditions generation ([easy](https://github.com/molgenis/molgenis-emx2/blob/75988d93e36b1814edb3ec25ef35d3184a671ad7/backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlQuery.java#L488)? I am not sure data is already joined?)

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/TriggerBuilderHelpers.java (hunk 1 above)
Comment (connoratrug) [reply backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/TriggerBuilderHelpers.java:13]: 100 % agree , can you suggest a DSL version for this ?

**Comment by connoratrug** [comment]
> it now also includes 'parent' and 'child' because that is included in the ontology search right? is that too eager?
> 
> Maybe this is more a comment about that we need to change the search index for ontology table than on this PR?
> 
> Meanwhile, I couldn't get it to work. => In pet store I annotated 'Tag=red' but then I didn't find results based on its label/code/etc
> 


> I suppose that is because I changed the 'color' label after the pet records were updated. So then I edited one of the pets and saved.
> 

> So we could accept this limitation but we might need to have some reindexing feature (hard) OR move the logic form the trigger to the SqlQuery -> search conditions generation ([easy](https://github.com/molgenis/molgenis-emx2/blob/75988d93e36b1814edb3ec25ef35d3184a671ad7/backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlQuery.java#L488)? I am not sure data is already joined?)

don't think moving to search query expansion is going to scale well, the general idea of updating the indexes on create/update time is the way to go for most ( and our ) use cases ( maybe need to optimise but bulk stuff ). We could improve this by having the trigger chain up te related fields, but would suggest leaving that for a separate pr ( again maybe this needs to be done lazy to avoid reindex chains on bulk insert/update, would need a good timing test to work/evaluate this).

**Comment by mswertz** [comment]
I think this issue is more severe that you think. I expect a lot of complaints of people not able to find stuff they expect because of updates. So we indeed then need this trigger to be implemented (will be of similar complexity as for the refback update feature we have). Meanwhile I am also not convinced that the performance hit is that bad, it even might be faster because one first filters the ontology terms and then the foreign key join will be using highly optimized index as compared to having to do a full table scan. (assuming the ontology table has less records than the data, which of course if not true for clinical ontologies). Is there a way we can assess the performance hit easily? Maybe by using a bit of '[explain analyze](https://www.postgresql.org/docs/current/sql-explain.html)' on the generated SQL queries?

**Comment by mswertz** [comment]
P.S. sonar test coverage is also not satisfied.

**Comment by connoratrug** [comment]
> P.S. sonar test coverage is also not satisfied.

yes any idea of how to test the migration ? , this is why i suggest real unit test as base coverage, and using dependency inject to write testable code

**Comment by connoratrug** [comment]
query expansion it it then

**Pull request closed without merging**